<table width = "100%">
  <tr style="background-color:white;">
    <!-- QWorld Logo -->
    <td style="text-align:left;width:200px;"> 
        <img src="../images/QWorld.png"> </td>
    <td style="text-align:right;vertical-align:bottom;font-size:16px;"> 
        Prepared by <a href="https://gitlab.com/pjr1363" target="_blank"> Paul Joseph Robin </a> and Özlem Salehi</td>
    </tr> 
 </table>
 
<hr>

# <font color="blue"> Solutions for </font> Ising Model

<a id="Task1"></a>
### Task 1

Given the Ising Model with the following properites, write a function to calculate the energy for different spin assignments and use that function to find the lowest energy state.

- There are 3 spins $s_0, s_1, s_2$.
- $h_0=4, h_1=2, h_2=-6$.
- $J_{0,1}=3, J_{0,2}=-1.3, J_{1,2}=2$

Your function should take as input the values for $s_0,s_1,s_2$ and return the energy.

### Solution

In [8]:
def calc_energy(s_0,s_1,s_2):
    return 3*s_0*s_1 + -1.3*s_0*s_2 + 2*s_1*s_2 + 4*s_0 + 2*s_1 + -6*s_2 

In [9]:
configurations = [[-1,-1,-1], [-1,-1,1],[-1,1,-1], [-1,1,1],[1,-1,-1], [1,-1,1],[1,1,-1], [1,1,1]]

conf_dict = {}
lowest_e = 1000
ground_state = []
for c in configurations:
    e = calc_energy(c[0],c[1],c[2])
    print("Energy for spin configuration",c,": ", e)
    if e < lowest_e:
        lowest_e = e
        ground_state = c
print("Ground state: ", ground_state)
print("Lowest energy: ", lowest_e)

Energy for spin configuration [-1, -1, -1] :  3.7
Energy for spin configuration [-1, -1, 1] :  -9.7
Energy for spin configuration [-1, 1, -1] :  -2.3000000000000007
Energy for spin configuration [-1, 1, 1] :  -7.7
Energy for spin configuration [1, -1, -1] :  8.3
Energy for spin configuration [1, -1, 1] :  -10.3
Energy for spin configuration [1, 1, -1] :  14.3
Energy for spin configuration [1, 1, 1] :  3.6999999999999993
Ground state:  [1, -1, 1]
Lowest energy:  -10.3


$s_0=1, s_1=-1, s_2=1$ is the lowest energy configuration corresponding to ground state, with energy -10.3. 

<a id="Task2"></a>

### Task 2

Construct an Ising model for the number partitioning problem.

Hint: Use $s_i$ to decide the group each number belongs to.

### Solution

Let $m_1$ and $m_2$ be the sum of the elements in group 1 and group 2 respectively.

We would like to have $m_1=m_2$. We can express this as an optimization problem where the goal is to minimize $(m_1-m_2)^2$.

We will use a similar idea as in Max-Cut so that 
$$s_{i}=
\left\{
\begin{array}{ll} 
      1, & \text{if number $n_i$ is in group 1} \\
      -1, & \text{if number $n_i$ is in group 2} \\
\end{array}
\right.$$


With this definition of $s_i$, we can express the Ising model as follows:

$$ \left( \sum_{i=1}^{N} n_i s_i \right)^2 .$$ 

Note that the summation is 0 iff two sets have the same sum and positive otherwise.